bep14.go

```diff
@@ -0,0 +1,476 @@
+		if err := p.SetMulticastInterface(iface); err != nil {
+			log.Printf("Set multicast interface error: %v\n", err)
+			return err
+		}
+	}
+	if m.network == "upd6" {
+		p := ipv6.NewPacketConn(m.mcPublisher)
+		if err := p.SetMulticastInterface(iface); err != nil {
+			log.Printf("Set multicast interface error: %v\n", err)
+			return err
+		}
+	}
+	return nil
+}
+
+func sourceUdpAddress(iface *net.Interface, network string) (*net.UDPAddr, error) {
+	addrs, err := iface.Addrs()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, addr := range addrs {
+		ipNet, ok := addr.(*net.IPNet)
+		if !ok {
+			continue
+		}
+
+		ip := ipNet.IP
+
+		// Skip loopback and link-local addresses
+		if ip.IsLoopback() || ip.IsLinkLocalUnicast() {
+			continue
+		}
+
+		switch network {
+		case "udp4":
+			if ip.To4() != nil {
+				return &net.UDPAddr{IP: ip}, nil
+			}
+		case "udp6":
+			if ip.To4() == nil && ip.To16() != nil {
+				return &net.UDPAddr{IP: ip}, nil
+			}
+		default:
+			return nil, errors.New("invalid network type, must be 'udp4' or 'udp6'")
+		}
+	}
+
+	return nil, errors.New("no suitable IP address found")
+}
+
+func lpdConnNew(network string, host string, lpd *LPDServer, config LocalServiceDiscoveryConfig) *LPDConn {
+	m := &LPDConn{}
+
+	m.lpd = lpd
+	m.network = network
+	m.host = host
+
+	var err error
+
+	m.addr, err = net.ResolveUDPAddr(m.network, m.host)
+	if err != nil {
+		log.Println("LPD unable to start", err)
+		return nil
+	}
+	m.mcListener, err = net.ListenMulticastUDP(m.network, nil, m.addr)
+	if err != nil {
+		log.Println("LPD unable to start", err)
+		return nil
+	}
+
+	m.mcPublisher, err = net.DialUDP(network, nil, m.addr)
+	if err != nil {
+		log.Println("Error dialing UDP:", err)
+		return nil
+	}
+
+	if config.Ifi != "" {
+		iface, err := net.InterfaceByName(config.Ifi)
+		if err != nil {
+			log.Printf("Interface error: %v\n", err)
+			return nil
+		}
+		sourceUdpAddress, err := sourceUdpAddress(iface, network)
+		if err != nil {
+			log.Printf("could not get source udp address: %v\n", err)
+			return nil
+		}
+		m.mcPublisher, err = net.DialUDP(network, sourceUdpAddress, m.addr)
+		if err != nil {
+			log.Println("Error dialing multicast interface:", err)
+			return nil
+		}
+		err = setMulticastInterface(m, iface)
+		if err != nil {
+			log.Println("Error setting multicast interface:", err)
+			return nil
+		}
+	} else {
+		m.mcPublisher, err = net.DialUDP(network, nil, m.addr)
+		if err != nil {
+			log.Println("Error dialing multicast interface:", err)
+			return nil
+		}
+	}
+
+	return m
+}
+
+func (m *LPDConn) receiver(client *Client) {
+	for {
+		m.lpd.mu.RLock()
+		if m.closed {
+			m.lpd.mu.RUnlock()
+			return
+		}
+		m.lpd.mu.RUnlock()
+
+		buf := make([]byte, 2000)
+		_, from, err := m.mcListener.ReadFromUDP(buf)
+		if err != nil {
+			log.Println("receiver", err)
+			continue
+		}
+
+		req, err := http.ReadRequest(bufio.NewReader(bytes.NewReader(buf)))
+		if err != nil {
+			log.Println("receiver", err)
```

Comment:
There is a defacto logging library in this repo. This message is going to be very hard to track down